File: airi-web/src/api/experiments.ts
import { http } from './client'
import type {
  DatasetSnapshotInfo,
  ExperimentReport,
  ExperimentSpec,
  LabelDefinition,
} from '../types/airi'

/**
 * Reads the seeded synthetic snapshot + label definition (see
 * scripts/seed_demo.py). Read-only on purpose: the browser never fabricates
 * dataset checksums. `name` pins the seeded demo snapshot so the later
 * temporal-slice snapshots do not shadow it.
 */
export async function fetchLatestDataset(name?: string): Promise<DatasetSnapshotInfo | null> {
  const { data } = await http.get<DatasetSnapshotInfo | null>('/api/v1/datasets/latest', {
    params: name ? { name } : undefined,
  })
  return data
}

/** `name` pins the seeded demo label for the same reason as the dataset. */
export async function fetchLatestLabel(name?: string): Promise<LabelDefinition | null> {
  const { data } = await http.get<LabelDefinition | null>('/api/v1/labels/latest', {
    params: name ? { name } : undefined,
  })
  return data
}

export async function createExperiment(payload: ExperimentSpec): Promise<{ experiment_spec_id: string }> {
  const { data } = await http.post('/api/v1/experiments', payload)
  return data
}

export async function runExperiment(specId: string): Promise<ExperimentReport> {
  const { data } = await http.post<ExperimentReport>(`/api/v1/experiments/${specId}/run`)
  return data
}

export async function fetchExperimentRun(runId: string): Promise<ExperimentReport> {
  const { data } = await http.get<ExperimentReport>(`/api/v1/experiments/runs/${runId}`)
  return data
}

File: src/airi/api/experiments.py
from fastapi import APIRouter, Depends, Request
from sqlalchemy.orm import Session

from airi.core.schemas import StrictSchema
from airi.experiments.models import (
    DatasetSnapshot,
    ExperimentReport,
    ExperimentSpec,
    LabelDefinition,
)
from airi.experiments.service import ExperimentService
from airi.infrastructure.database import get_session

router = APIRouter(prefix="/api/v1", tags=["experiments"])


class RunRequest(StrictSchema):
    """Execution accepts no SQL or other caller-controlled fields."""


def service(request: Request, session: Session = Depends(get_session)):
    return ExperimentService(session, request.app.state.query_executor, request.app.state.settings)


@router.post("/datasets", response_model=DatasetSnapshot, status_code=201)
def dataset(payload: DatasetSnapshot, app: ExperimentService = Depends(service)):
    return app.register_dataset(payload)


@router.post("/labels", response_model=LabelDefinition, status_code=201)
def label(payload: LabelDefinition, app: ExperimentService = Depends(service)):
    return app.register_label(payload)


@router.get("/datasets/latest", response_model=DatasetSnapshot | None)
def latest_dataset(name: str | None = None, app: ExperimentService = Depends(service)):
    """Read-only: lets the Web demo reuse the seeded synthetic snapshot.

    `name` pins the seeded demo snapshot so temporal-slice datasets do not
    shadow it.
    """
    return app.latest_dataset(name)


@router.get("/labels/latest", response_model=LabelDefinition | None)
def latest_label(name: str | None = None, app: ExperimentService = Depends(service)):
    return app.latest_label(name)


@router.post("/experiments", response_model=ExperimentSpec, status_code=201)
def create(payload: ExperimentSpec, app: ExperimentService = Depends(service)):
    return app.create(payload)


@router.post("/experiments/{spec_id}/run", response_model=ExperimentReport)
def run(
    spec_id: str,
    request: Request,
    payload: RunRequest | None = None,
    app: ExperimentService = Depends(service),
):
    return app.run(spec_id, request.state.request_id)


@router.get("/experiments/runs/{run_id}", response_model=ExperimentReport)
def report(run_id: str, app: ExperimentService = Depends(service)):
    return app.report(run_id)
